### PR TITLE
Add `python_compiler` utilities for ease of use and testing

### PR DESCRIPTION
**Pull request by mudit2812**
* Add `inline_module` function that takes two modules as input, and inlines the body of the first module into the second module.
* Add `inline_jit_to_module` function that takes a `jax.jit`-ed function and xDSL module as inlines the xDSL module corresponding to the jitted function into the user provided module. This function doesn't return anything; it modifies the module in-place.
* Add `xdsl_from_qjit` function, which takes a QJIT-ed function as input and returns a wrapper that, when called, returns an xDSL module.
[sc-93673] [sc-93672] [sc-93735]

### CODE REVIEW

File: tests/python_compiler/conftest.py [hunk 1]
```diff
@@ -43,7 +61,48 @@ def _run_filecheck(program_str, xdsl_module):
...
+        ctx.load_dialect(Quantum)
+
+        xmod = Parser(ctx, mod_generic).parse_module()
+        return xmod
+
+    return wrapper
+
+
+@pytest.fixture(scope="function")
+def from_qjit():
```
Comment (mudit2812) [inline tests/python_compiler/conftest.py]: Instead of a fixture, this could potentially be part of the `qml.compiler.python_compiler` submodule.

File: pennylane/compiler/python_compiler/jax_utils.py [hunk 2]
```diff
@@ -125,12 +119,25 @@ def wrapper(*_, **__):
...
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> xfunc.FuncOp:
+        xmod = _xdsl_module_inline(func, *args, **kwargs)
+        funcOp = xmod.regions[0].blocks[0].first_op
+        # We create a clone because the original FuncOp is attached to xmod's region
+        funcOp = funcOp.clone()
+        funcOp.properties["sym_name"] = xbuiltin.StringAttr(f"{func.__name__}")
+        return funcOp
+
+    return wrapper
```
Comment (erick-xanadu) [inline pennylane/compiler/python_compiler/jax_utils.py]: This is in general wrong, you should return all functions. Otherwise the function that you are extracting may be calling other functions. This is why I opted not to get the function directly. You may leave it here, but I recommend adding an assertion that there is only one function or change the return signature and only use this for internal developers.

Comment (mudit2812) [reply pennylane/compiler/python_compiler/jax_utils.py]: Oof you are correct. Something to think about once I have some more time 😅

File: tests/python_compiler/conftest.py [hunk 3]
```diff
@@ -43,7 +61,48 @@ def _run_filecheck(program_str, xdsl_module):
...
+    def wrapper(*args, **kwargs):
+        _ = qjit_fn(*args, **kwargs)
+        mod = qjit_fn.mlir
+        mod_generic = _quantum_opt(
+            _quantum_opt(
+                ("--pass-pipeline", "builtin.module(canonicalize)"),
+                "-mlir-print-op-generic",
+                stdin=mod,
+            )
+        )
```
Comment (erick-xanadu) [inline tests/python_compiler/conftest.py]: I think this could be simplified a bit, but I am not sure yet how. :thinking: qjit_fn also has `.mlir_module` which may be printed in generic format.

File: pennylane/compiler/python_compiler/jax_utils.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/compiler/python_compiler/jax_utils.py]: I removed this function and replaced it with one that takes an xDSL module as input and inlines a jitted function's IR into it.

File: tests/python_compiler/conftest.py (hunk 3 above)
Comment (mudit2812) [reply tests/python_compiler/conftest.py]: Updated

File: pennylane/compiler/python_compiler/jax_utils.py [hunk 4]
```diff
@@ -125,12 +120,38 @@ def wrapper(*_, **__):
...
+
+
+def copy_jit_to_module(func: JaxJittedFunction, mod: xbuiltin.ModuleOp, *args, **kwargs) -> None:
+    """Inline a ``jax.jit``-ed Python function to an xDSL module."""
+    func_mod = _xdsl_module_inline(func, *args, **kwargs)
+
+    for op in func_mod.body.ops:
+        if isinstance(op, xfunc.FuncOp) and op.sym_name.data == "main":
+            op.properties["sym_name"] = xbuiltin.StringAttr(func.__name__)
+            break
```
Comment (erick-xanadu) [inline pennylane/compiler/python_compiler/jax_utils.py]: I think there's a way to look the symbol name in the symbol table. 

```
        funcOp = SymbolTable.lookup_symbol(module, function_name)
```

Comment (erick-xanadu) [reply pennylane/compiler/python_compiler/jax_utils.py]: Maybe as a TODO something to add to xDSL would be to add a rename operation to SymbolTable that would be able to update all callsites to the renamed function.

https://github.com/xdslproject/xdsl/blob/c122d1ce5f7854ffc868a4e7bacecb669b81e8c8/xdsl/traits.py#L327

Comment (erick-xanadu) [reply pennylane/compiler/python_compiler/jax_utils.py]: But I think you can use insert or update to insert the new functions. Just be careful if they already exist as they will override the previous definition.

Comment (mudit2812) [reply pennylane/compiler/python_compiler/jax_utils.py]: Thanks for finding this. Updated.

File: pennylane/compiler/python_compiler/jax_utils.py [hunk 5]
```diff
@@ -125,12 +121,36 @@ def wrapper(*_, **__):
...
+
+def copy_jit_to_module(func: JaxJittedFunction, mod: xbuiltin.ModuleOp, *args, **kwargs) -> None:
+    """Inline a ``jax.jit``-ed Python function to an xDSL module. The inlined body is appended
+    to the end of ``mod``."""
+    func_mod = _xdsl_module_inline(func, *args, **kwargs)
+    main_func = xSymbolTable.lookup_symbol(func_mod, "main")
+    main_func.properties["sym_name"] = xbuiltin.StringAttr(func.__name__)
+
+    cloned_ops = tuple(op.clone() for op in func_mod.body.ops)
+    mod.body.blocks[0].add_ops(cloned_ops)
```
Comment (erick-xanadu) [inline pennylane/compiler/python_compiler/jax_utils.py]:
```suggestion
    for func in func_mod:
        xSymbolTable.insert_or_update(mod, func)
```

Try this?

Comment (erick-xanadu) [reply pennylane/compiler/python_compiler/jax_utils.py]: Do you need to clone?

Comment (mudit2812) [reply pennylane/compiler/python_compiler/jax_utils.py]: Yes, I tried without cloning and got the usual "Can't insert op that is already attached to a block" error. I could just use `op.detach()` instead of `op.clone()`, but `clone()` seems safer. Now that I've updated the implementation to use the new `inline_module` utility I added, I also cannot assume that the original module will not get reused.

**Review by erick-xanadu — APPROVED** [review]
